streams.cpp

```diff
@@ -258,3 +279,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
     ST::hex_decode(value, result, resultLen);
     return Blob::Steal(result, resultLen);
 }
+
+DS::EncryptedStream::EncryptedStream(
+    DS::Stream* base, DS::EncryptedStream::Mode mode,
+    std::optional<DS::EncryptedStream::Type> type, const uint32_t* keys
+) : m_base(base), m_buffer(), m_key(), m_pos(), m_size(),
+    m_type(type.has_value() ? type.value() : DS::EncryptedStream::Type::e_tea),
+    m_mode(mode)
+{
+    DS_ASSERT(base != nullptr);
+    DS_ASSERT(base->tell() == 0);
+
+    static constexpr uint32_t kTeaKey[] {
+        0x6c0a5452,
+        0x03827d0f,
+        0x3a170b92,
+        0x16db7fc2
+    };
+    if (keys) {
+        memcpy(m_key, keys, sizeof(m_key));
+    } else {
+        static_assert(sizeof(kTeaKey) == sizeof(m_key));
+        memcpy(m_key, kTeaKey, sizeof(m_key));
+    }
+
+    uint8_t header[16]{};
+    DS_ASSERT(!(!type.has_value() && mode == Mode::e_write));
+    switch (mode) {
+    case Mode::e_read:
+        base->readBytes(header, sizeof(header));
+        if (memcmp(header, "whatdoyousee", 12) == 0)
+            m_type = DS::EncryptedStream::Type::e_tea;
+        else if (memcmp(header, "BriceIsSmart", 12) == 0)
+            m_type = DS::EncryptedStream::Type::e_tea;
+        else if (memcmp(header, "notthedroids", 12) == 0)
+            m_type = DS::EncryptedStream::Type::e_xxtea;
+        else
+            throw DS::FileIOException("Unknown EncryptedString magic");
+        DS_ASSERT(!type.has_value() || type.value() == m_type);
+        m_size = reinterpret_cast<uint32_t*>(header)[3];
+        break;
+
+    case Mode::e_write:
+        // Write out some temporary junk for the header for now.
+        uint8_t header[16]{};
+        base->writeBytes(header, sizeof(header));
+        break;
+    }
+}
+
+DS::EncryptedStream::~EncryptedStream()
+{
+    if (m_mode == Mode::e_write) {
+        if (m_pos % sizeof(m_buffer) != 0)
+            cryptFlush();
+        m_base->seek(0, SEEK_SET);
+        switch (m_type) {
+            case Type::e_xxtea:
+                m_base->writeBytes("notthedroids", 12);
+                break;
+            case Type::e_tea:
+                m_base->writeBytes("whatdoyousee", 12);
+                break;
+        }
+        m_base->write<uint32_t>(m_size);
+    }
```

Comment:
I would probably prefer this to be in a separate function that is then called by the destructor...  Otherwise, the destructor is the only way to finalize the stream header.